Add after import event

koertho opened the pull request:
This PR add a new event to adjust calendar events after the csv import. 

I use it to add images added as attachment to the calendar event.

I build it as symfony event instead of a hook to use the more current standard.

Example usage (just a first shoot): 

```php
<?php

namespace App\EventListener;

use Ausi\SlugGenerator\SlugGenerator;
use Ausi\SlugGenerator\SlugOptions;
use Cgoit\ContaoCalendarIcalBundle\Event\AfterImportItemEvent;
use Contao\CoreBundle\Filesystem\VirtualFilesystemInterface;
use Symfony\Component\DependencyInjection\ParameterBag\ParameterBagInterface;
use Symfony\Component\EventDispatcher\Attribute\AsEventListener;

#[AsEventListener(AfterImportItemEvent::class)]
class ImportIcalEventListener
{
    public function __construct(
        private readonly VirtualFilesystemInterface $filesStorage,
        private readonly ParameterBagInterface $parameterBag,
    ) {
    }

    public function __invoke(AfterImportItemEvent $event)
    {
        if (!$event->vevent->getAttach() || $event->calendarEventsModel->singleSRC) {
            return;
        }

        $attachment = $event->vevent->getAttach(0, true);

        if ($attachment && str_starts_with($attachment->getValue(), 'https://drive.google.com/file/d/')) {
            $filename = $attachment->getParams('FILENAME');
            if (!$filename) {
                return;
            }
            $fileInfo = pathinfo($filename);
            $filename = (new SlugGenerator((new SlugOptions())
                    ->setValidChars('a-z0-9')
                    ->setDelimiter('-')
                ))->generate($fileInfo['filename'])
                . '.' . $fileInfo['extension'];

            $id       = substr($attachment->getValue(), 32);
            $fileUrl = 'https://drive.google.com/uc?export=download&id=' . $id;

            $year = date('Y', $event->calendarEventsModel->startDate);

            $filePath = 'media/veranstaltungen/' . $year . '/import/' . $event->calendarEventsModel->alias . '_' . $filename;
            try {
                $this->filesStorage->writeStream(
                    $filePath,
                    fopen($fileUrl, 'r'),
                );

            } catch (\Exception $exception) {
                return;
            }

            $fileItem = $this->filesStorage->get($filePath);

            if ($fileItem) {
                $event->calendarEventsModel->addImage = '1';
                $event->calendarEventsModel->singleSRC = $fileItem->getUuid()->toBinary();
                $event->calendarEventsModel->save();
            }

        }
    }
}
```

cgoIT commented:
First of all thx for this contribution.

I just need to understand why you implemented this as a Symfony Event and not as a hook. You are talking about a "new standard". Can you give me a link to a reference or something like that where contao proposed Symfony Events as a new standard instead of hooks. I'm asking because we already have some HOOKs and at least in my POV events and hooks are to slightly different things.

And last but not least if we should put this into a new version we would need at least one [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/).

koertho commented:
Hi, I could not find an offical statemant, so I asked in slack: https://contao.slack.com/archives/CJT61P8R1/p1745817378823379 :)

But from my point of view, where contao is shifting it's standard more and more to symfony, it's only logical. New "event" are mostly implemented as symfony event in the contao core while some hooks already got deprecated. Also is the event system of symfony a way better implementation then the contao hooks.

fritzmg commented:
That's right, hooks are a legacy and outdated concept from Contao 2 and 3. If you are employing event-driven architecture within your own code, you should of course use the Symfony Event Dispatcher ever since Contao 4. (Some extensions even used this prior to Contao 4.)

koertho commented:
I adjusted a variable name and added an empty commit trying to make an conventional commit, hope that's right. But my suggestion for the future is to use the github squash feature, than you can set the commit message by yourself and have a cleaner history :)

cgoIT commented:
Ok. I have no access to this slack, but I'll approve the MR now. Thx for your contribution.

koertho commented:
Thanks! Thats the offical contao slack, very helpfull :)
